### ISSUE TITLE
Move dask-pytorch under dask org

### ISSUE THREAD

**Issue by jsignell** [dask/community#111]
This came up in https://github.com/dask/community/issues/103#issuecomment-731152084 but I wanted to make sure it got its own discussion. 

@hhuuggoo and @skirmer recently released dask-pytorch: https://github.com/saturncloud/dask-pytorch and @TomAugspurger suggested that it should move into the dask org. 

Is that ok with folks?

**Comment by martindurant** [comment]
+1

**Comment by mrocklin** [comment]
Historically we incubate things outside of the main org, and then once they're mature enough we bring them in.  I suggest that some Dask folks take a look at the implementation, see how things look, and make a judgement after that assessment.

The current implementation looks simple (which is encouraging) but also sometimes also specific (referring specifically to the `S3ImageFolder` class).  It may or may not be the preferred medium-range architecture.

To throw out an example, this implementation currently includes a single `S3ImageFolder` dataset class.  This doesn't seem like general purpose functionality, or scope that we would want to support, but instead seems like the kind of thing that was useful to satisfy the needs of a specific client.  Things like this might be better either generalized, or removed out to some specific  `saturn-dask-pytorch` utils package.  This may require a longer conversation.

I think that there are two general questions: 

1.  What is the scope/architecture of the project that Dask wants to maintain?
2.  How does Saturn want to handle their more customer-specific work that may not fit in a general purpose package?

**Comment by hhuuggoo** [comment]
Happy to keep incubating it until the appropriate time.

w.r.t the dataset - that's a good point.  Our client projects are actually all around video processing, the `S3ImageFolder` was only created so we could show a simple example of large scale parallel training.  Using `dask-pytorch` doesn't actually require any special datasets, most of the existing PyTorch datasets work with files on disk, and so if your dask workers already have access to these files, those classes work just fine.  

We should think about lifting out datasets completely - possibly into another library which probably would not even depend on dask.

The nice thing about this is, that then `dask-pytorch` would be close to complete.  The only other thing I was thinking about adding were  other result mechanisms that do not use the dask pub-sub pattern (intermediate results directly to s3/azure blob/gcs) , and possibly some tooling around displaying these same results in tensorboard.  After that it could probably just be in maintenance mode.  Famous last words?

**Comment by mrocklin** [comment]
I spent about an hour this morning learning about PyTorch DDP.  I came up with a similar implemenation in this [gist](https://gist.github.com/mrocklin/f2df45a699f0381384af26384c324589) (should be runnable from a laptop, disclaimer I can't speak for the accuracy of the actual training process).  I worked from [this tutorial](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html) which, looking at your code now, I'm guessing you all did as well.

Some general feedback / requests after looking at this project and this topic more closely:

### Thoughts on this implementation

1.  It would be useful to have a test or example that could be run on a single machine.  All of the tests currently mock things out, and don't test actual PyTorch usage.  This makes it both hard for new users to try things out, and results in slightly less confidence that things work
2.  I suspect that this is because some of this is still fairly hard-coded to one specific use case.  For example, you've hard-coded the nccl backend, which I suspect would only work if nccl is installed and if GPUs are present.  I would encourage you all to make this configurable
3.  More broadly there are a bunch of other keywords to `dist.init_process_group` that appear to now be hidden.
4.  This project currently assumes that the futures will be evenly spread out, and so would fail if *anything* else is running on the same cluster at the same time.  I recommend using worker restrictions
5.  What happens if this is run on Windows?  (my solution doesn't handle this well either)
6.  It looks like we destroy all process groups when we clean up (I'm not sure how common multiple sessions would be, so this probably doesn't matter)

### Thoughts on this topic more broadly

I wonder if the name "dask-pytorch" is a bit too grand for this work.  Maybe instead "dask-pytorch-ddp" would be better.   There are *many* different approaches to distributing PyTorch.  For example two others in this space would be to support or back Horovod, or to build our own distributed training with pubsub/actors/whatever.  Beyond that, there are possibly many other things that one might want to do between PyTorch and Dask that people might expect this package to do.  

Alternatively, I wonder if perhaps a better approach would be to submit a PR to dask-ml and have this live there for a bit until there is enough substance here to warrant a separate project with distinct docs and such.

There is some history here.  A long time ago I made dask-tensorflow, largely at the request of Anaconda's Marketing department.  It was similar in nature to this project.  It made it easy to set up TensorFlow from Dask, and then showed a single workflow to use them together.  It got people excited and generated some marketing buzz, but ended up resulting in a lot of disappointed users who expected it to do something very different.  In hindsight, I regret publishing that project and we eventually discontinued it.  I would encourage us not to make the same mistake here.

I would love for a dask-pytorch package to exist someday.  My guess is that that will come after working with several different groups and finding very common patterns among them.  When thinking about design I would also recommend that people look at packages in other ecosystems like Spark to get a sense for what workflows others have found to be broadly pragmatic.

**Comment by TomAugspurger** [comment]
I haven’t played with dask-pytorch yet, but as we go through this process I’d encourage people to think about generalizing things

1. What requirements do we have for including a `dask-foo` project in the dask org
2. What does a project being in the dask org mean (in terms of maintenance, supported versions, code of conduct, …), if anything.
3. What policy do we have on `dask-foo` projects that aren’t in the Dask org (NumPy has https://numpy.org/neps/nep-0036-fair-play.html <https://numpy.org/neps/nep-0036-fair-play.html> which lays this out).

The goal being that we don’t have to make arbitrary choices next time.

FWIW, I was also thinking in terms of avoiding another “dask-tensorflow” issue (again, I haven’t even installed dask-pytorch yet; comparing the two probably isn’t fair).

**Comment by hhuuggoo** [comment]
Thanks for all the feedback.  I think this is a good conversation, and will likely continue for some time.  I need to incubate this package by iterating on it with users.  Because of the feedback my current plan is to rename it to `dask-pytorch-ddp` (in the saturncloud github org).  Are there any objections to this? Please let me know.

w.r.t dask-ml, fwiw I'm against putting pytorch stuff in there only because dask-ml has a fairly clear scope (dask for sklearn) and I think adding pytorch will make it less cohesive and less complete.

**Comment by mrocklin** [comment]
> w.r.t dask-ml, fwiw I'm against putting pytorch stuff in there only because dask-ml has a fairly clear scope (dask for sklearn) and I think adding pytorch will make it less cohesive and less complete.

Dask-ML was actually created as a namespace package for all of the different Dask ML libraries out there (glm, hyperparameter-optimization, xgboost, tensorflow, etc..).  I think that centralizing was generally a good choice and made things more cohesive and discoverable.

> Because of the feedback my current plan is to rename it to dask-pytorch-ddp (in the saturncloud github org). Are there any objections to this? Please let me know.

People do run their own dask-foo libraries.  Generally as they succeed we encourage folks to upstream (for a similarly scoped example, see jrbourbeau/dask-optuna which is now trying to upstream into optuna core).  `<puts on for-profit hat>` putting things into a for-profit github org makes it less likely that cross-company collaboration (as above) will occur.  I think that moving things like this into the commons is good in order to avoid for-profit conflicts.  I totally get that having a dask-pytorch library in the saturncloud github org looks good though, so no pressure `</takes off for-profit hat>`.  

I may at some point solidify my version (which I think may be somewhat more robust) and push it up as a PR into dask-ml and then ask some of the maintainers here to provide feedback.  This *seems* like the right approach to me, but I'm open to suggestions.

**Comment by hhuuggoo** [comment]
I meant I'm planning on incubating `dask-pytorch-ddp` in the saturncloud gh org in the short term as it develops.  I am +2 on migrating it to dask org, and +1 on a dask-ml PR.  I personally think all of ML in dask-ml is too broad.  I would be -1 on leaving it in the saturncloud org long term - for the reasons you stated.
